syntax = "proto2";
package buffers.resources;

import "options.proto";

message Object {
  message EgmEvent {  // RM = Rename me.
    optional string id = 1;
    repeated string arguments = 2;
    optional string code = 3;
  }
  message LegacyEvent {
    optional int32 type = 1 [(gmx) = "eventtype"];
    oneof reference {
      int32 number = 2 [(gmx) = "enumb"];
      string name = 3 [(gmx) = "ename", (yyp_id) = "collisionObjectId"];
    }
    optional string code = 4 [(gmx) = "action"];
  }

  enum PhysicsShape { CIRCLE = 0; BOX = 1; SHAPE = 2; }

  message PhysicsShapePoint {
    optional int32 x = 1 [(gmx) = "GMX_SPLIT/0"];
    optional int32 y = 2 [(gmx) = "GMX_SPLIT/1"];
  }

  message ObjectPhysicsSettings {
    optional bool use_physics = 1 [(gmx) = "PhysicsObject", (yyp) = "physicsObject"];
    optional bool phy_sensor = 2 [(gmx) = "PhysicsObjectSensor", (yyp) = "physicsSensor"];
    optional PhysicsShape phy_shape_type = 3 [(gmx) = "PhysicsObjectShape", (yyp) = "physicsShape"];
    optional double phy_density = 4 [(gmx) = "PhysicsObjectDensity", (yyp) = "physicsDensity"];
    optional double phy_restitution = 5 [(gmx) = "PhysicsObjectRestitution", (yyp) = "physicsRestitution"];
    optional int32 phy_group = 6 [(gmx) = "PhysicsObjectGroup", (yyp) = "physicsGroup"];
    optional double phy_linear_damping = 7 [(gmx) = "PhysicsObjectLinearDamping", (yyp) = "physicsLinearDamping"];
    optional double phy_angular_damping = 8 [(gmx) = "PhysicsObjectAngularDamping", (yyp) = "physicsAngularDamping"];
    optional double phy_friction = 9 [(gmx) = "PhysicsObjectFriction", (yyp) = "physicsFriction"];
    optional bool phy_awake = 10 [(gmx) = "PhysicsObjectAwake", (yyp) = "physicsAwake"];
    optional bool phy_kinematic = 11 [(gmx) = "PhysicsObjectKinematic", (yyp) = "physicsKinematic"];
    repeated PhysicsShapePoint phy_shape = 12 [(gmx) = "PhysicsShapePoints/point", (yyp) = "physicsShapePoints"];
  }

  optional int32 id = 2 [(gmx) = "GMX_DEPRECATED"];

  optional string parent_name = 3 [(resource_ref)="object", (gmx) = "parentName", (yyp_id) = "parentObjectId"];
  optional string sprite_name = 4 [(resource_ref)="sprite", (gmx) = "spriteName", (yyp_id) = "spriteId"];
  optional string mask_name = 5 [(resource_ref)="sprite", (gmx) = "maskName", (yyp_id) = "maskSpriteId"];

  optional int32 depth = 6 [(yyp) = "YYP_DEPRECATED"];
  optional bool solid = 7;
  optional bool visible = 8;
  optional bool persistent = 9;

  repeated LegacyEvent legacy_events = 10 [(gmx) = "events/event", (yyp) = "eventList"];
  repeated EgmEvent egm_events = 12;

  optional ObjectPhysicsSettings physics_settings = 11 [(gmx) = "EGM_NESTED"];
}
